File: pkg/connectors/discord.go
package connectors

import (
	"encoding/csv"
	"encoding/json"
	"fmt"
	"io"
	"os"
	"path/filepath"
	"sort"
	"strings"
	"time"

	"github.com/jschell12/replicateme/pkg/corpus"
)

// DiscordImportOptions controls Discord data export import.
type DiscordImportOptions struct {
	File  string // path to ZIP or extracted directory
	Since *time.Time
}

// ImportDiscord reads messages from a Discord data export.
func ImportDiscord(opts DiscordImportOptions) ([]corpus.RawMessage, error) {
	dir := opts.File

	if strings.HasSuffix(opts.File, ".zip") {
		dir = "/tmp/replicateme-discord-export"
		if err := extractZip(opts.File, dir); err != nil {
			return nil, fmt.Errorf("extract discord zip: %w", err)
		}
	}

	if _, err := os.Stat(dir); err != nil {
		return nil, fmt.Errorf("Discord export not found at %s", dir)
	}

	messagesDir := dir
	if _, err := os.Stat(filepath.Join(dir, "messages")); err == nil {
		messagesDir = filepath.Join(dir, "messages")
	}

	// load channel index
	channelNames := make(map[string]string)
	indexFile := filepath.Join(messagesDir, "index.json")
	if data, err := os.ReadFile(indexFile); err == nil {
		var index map[string]*string
		if json.Unmarshal(data, &index) == nil {
			for id, name := range index {
				if name != nil {
					channelNames[id] = *name
				}
			}
		}
	}

	var messages []corpus.RawMessage

	entries, err := os.ReadDir(messagesDir)
	if err != nil {
		return nil, err
	}

	for _, entry := range entries {
		if !entry.IsDir() {
			continue
		}
		channelID := entry.Name()
		csvPath := filepath.Join(messagesDir, channelID, "messages.csv")
		if _, err := os.Stat(csvPath); err != nil {
			continue
		}

		channelName := channelNames[channelID]
		if channelName == "" {
			channelName = channelID
		}

		f, err := os.Open(csvPath)
		if err != nil {
			continue
		}

		reader := csv.NewReader(f)
		reader.LazyQuotes = true

		// skip header
		if _, err := reader.Read(); err != nil {
			f.Close()
			continue
		}

		for {
			record, err := reader.Read()
			if err == io.EOF {
				break
			}
			if err != nil {
				continue
			}
			// columns: ID, Timestamp, Contents, Attachments
			if len(record) < 3 {
				continue
			}

			id := record[0]
			timestampStr := record[1]
			contents := record[2]

			if id == "" || contents == "" || timestampStr == "" {
				continue
			}

			ts, err := time.Parse(time.RFC3339, timestampStr)
			if err != nil {
				// try other formats
				ts, err = time.Parse("2006-01-02 15:04:05", timestampStr)
				if err != nil {
					continue
				}
			}

			if opts.Since != nil && ts.Before(*opts.Since) {
				continue
			}

			messages = append(messages, corpus.RawMessage{
				ID:         fmt.Sprintf("discord-%s", id),
				Text:       contents,
				Platform:   corpus.Discord,
				Timestamp:  ts.UTC(),
				IsFromUser: true,
				Metadata: map[string]any{
					"channel":   channelName,
					"channelId": channelID,
				},
			})
		}
		f.Close()
	}

	sort.Slice(messages, func(i, j int) bool {
		return messages[i].Timestamp.Before(messages[j].Timestamp)
	})
	return messages, nil
}

File: src/connectors/discord.ts
import { readFileSync, existsSync, readdirSync } from "fs";
import { join } from "path";
import { execSync } from "child_process";
import type { RawMessage } from "../corpus/types.js";

// Discord data export format:
// ZIP containing messages/ directory with subdirectories per channel.
// Each channel dir has messages.csv with columns:
// ID,Timestamp,Contents,Attachments
// Also has index.json mapping channel IDs to names.

export interface DiscordImportOptions {
  file: string; // path to ZIP or extracted directory
  since?: Date;
}

export function importDiscord(opts: DiscordImportOptions): RawMessage[] {
  let dir = opts.file;

  if (opts.file.endsWith(".zip")) {
    dir = "/tmp/replicateme-discord-export";
    execSync(`rm -rf ${dir} && mkdir -p ${dir} && unzip -q -o "${opts.file}" -d ${dir}`);
  }

  if (!existsSync(dir)) {
    throw new Error(`Discord export not found at ${dir}`);
  }

  const messagesDir = existsSync(join(dir, "messages"))
    ? join(dir, "messages")
    : dir;

  // load channel index if available
  const channelNames: Record<string, string> = {};
  const indexFile = join(messagesDir, "index.json");
  if (existsSync(indexFile)) {
    const index = JSON.parse(readFileSync(indexFile, "utf-8")) as Record<
      string,
      string | null
    >;
    for (const [id, name] of Object.entries(index)) {
      if (name) channelNames[id] = name;
    }
  }

  const messages: RawMessage[] = [];

  const channelDirs = readdirSync(messagesDir, { withFileTypes: true })
    .filter((d) => d.isDirectory())
    .map((d) => d.name);

  for (const channelId of channelDirs) {
    const csvPath = join(messagesDir, channelId, "messages.csv");
    if (!existsSync(csvPath)) continue;

    const channelName = channelNames[channelId] ?? channelId;
    const csv = readFileSync(csvPath, "utf-8");
    const rows = parseCSV(csv);

    for (const row of rows) {
      const [id, timestampStr, contents] = row;
      if (!id || !contents || !timestampStr) continue;

      const timestamp = new Date(timestampStr);
      if (isNaN(timestamp.getTime())) continue;
      if (opts.since && timestamp < opts.since) continue;

      messages.push({
        id: `discord-${id}`,
        text: contents,
        platform: "discord",
        timestamp,
        isFromUser: true, // discord export only contains your own messages
        metadata: {
          channel: channelName,
          channelId,
        },
      });
    }
  }

  messages.sort((a, b) => a.timestamp.getTime() - b.timestamp.getTime());
  return messages;
}

function parseCSV(content: string): string[][] {
  const rows: string[][] = [];
  const lines = content.split("\n");

  // skip header row
  for (let i = 1; i < lines.length; i++) {
    const line = lines[i]!.trim();
    if (!line) continue;
    rows.push(parseCSVLine(line));
  }

  return rows;
}

function parseCSVLine(line: string): string[] {
  const fields: string[] = [];
  let current = "";
  let inQuotes = false;

  for (let i = 0; i < line.length; i++) {
    const ch = line[i]!;
    if (inQuotes) {
      if (ch === '"') {
        if (i + 1 < line.length && line[i + 1] === '"') {
          current += '"';
          i++; // skip escaped quote
        } else {
          inQuotes = false;
        }
      } else {
        current += ch;
      }
    } else {
      if (ch === '"') {
        inQuotes = true;
      } else if (ch === ",") {
        fields.push(current);
        current = "";
      } else {
        current += ch;
      }
    }
  }
  fields.push(current);
  return fields;
}
